scrapy_autoextract/providers.py

```diff
@@ -1,114 +1,229 @@
+        set using the setting `CONCURRENT_REQUESTS_PER_DOMAIN`.
 
-def install():
-    """Register all providers for their respective provided classes."""
-    ArticleDataProvider.register()
-    ProductDataProvider.register()
+        By default the key is the request domain. Override it to change
+        the behavior.
+        """
+        return get_domain(request.url)
+
+    def get_filled_request(self,
+                           request: ScrapyRequest,
+                           provided_cls: AutoExtractData,
+                           should_request_html: bool) -> AutoExtractRequest:
+        """Return a filled request for AutoExtract"""
+        ae_request = AutoExtractRequest(
+            url=request.url,
+            pageType=provided_cls.page_type,
+        )
+        if should_request_html:
+            ae_request.extra = {self.html_query_attribute: True}
+        return ae_request
+
+    async def __call__(self,
+                       to_provide: Set[Callable],
+                       request: ScrapyRequest,
+                       stats: StatsCollector
+                       ) -> List:
+        """Make an AutoExtract request and build a Page Input of provided class
+        based on API response data.
+        """
+        if not self.aiohttp_session:
+            self.aiohttp_session = await self.create_aiohttp_session()
+        is_html_required = AutoExtractHtml in to_provide
+        to_provide -= {AutoExtractHtml}
+        is_extraction_required = bool(to_provide)
+        if is_html_required and not is_extraction_required:
+            # At least one request is required to get html
+            to_provide = {self.page_type_class_for_html}
+
+        instances = []
+        for idx, provided_cls in enumerate(to_provide):
+            if not issubclass(provided_cls, AutoExtractData):
+                raise RuntimeError(
+                    f"Unexpected {provided_cls} requested. Probably a bug in the provider "
+                    "or in scrapy-poet itself")
+            request_stats = AggStats()
+            is_first_request = idx == 0
+            page_type = provided_cls.page_type
+
+            def inc_stats(suffix, value=1, both=False):
+                stats.inc_value(f"autoextract/total{suffix}", value)
+                if both:
+                    stats.inc_value(f"autoextract/{page_type}{suffix}", value)
+
+            try:
+                # html is requested only a single time to save resources
+                should_request_html = is_html_required and is_first_request
+                slot = self.get_per_domain_concurrency_slot(request)
+                ae_request = self.get_filled_request(
+                    request,
+                    provided_cls,
+                    should_request_html
+                )
+                if not ae_request:
+                    raise ValueError("Unexpected empty return from method get_filled_request")
```

Comment:
Why is this needed, won't it fail at some other place if get_filled_request is buggy?